netbench/netbench-collector/src/main.rs

```diff
@@ -18,24 +36,363 @@ pub struct Args {
+    state: State,
+}
 
-    // fall back to the generic collector
-    generic::run(&args)?;
+impl From<State> for StateMessage {
+    fn from(value: State) -> StateMessage {
+        Self {
+            version: "18-04-2023".to_string(),
+            state: value,
+        }
+    }
+}
 
-    Ok(())
+impl TryFrom<u8> for State {
+    type Error = ();
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(State::NotReady),
+            1 => Ok(State::Ready),
+            2 => Ok(State::Running),
+            3 => Ok(State::Finished),
+            _ => Err(()),
+        }
+    }
+}
+
+impl From<State> for u8 {
+    fn from(value: State) -> u8 {
+        match value {
+            State::NotReady => 0,
+            State::Ready => 1,
+            State::Running => 2,
+            State::Finished => 3,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct StateTracker {
+    current_state: Arc<AtomicU8>,
+    location: String,
+    other_location: String,
+    verbose: bool,
+}
+
+macro_rules! log_polling {
+    ($verbose:ident, $($arg:tt)*) => {
+        if $verbose {
+            println!("[POLLING] {}", format!($($arg)*));
+        }
+    }
+}
+
+macro_rules! log_serving {
+    ($verbose:ident, $($arg:tt)*) => {
+        if $verbose {
+            println!("[SERVING] {}", format!($($arg)*));
+        }
+    }
+}
+
+impl StateTracker {
+    fn store(&mut self, value: State) {
+        if self.verbose {
+            println!("[LOCAL STATE] Updating local state to {:?}", value);
+        }
+        self.current_state.store(value.into(), Ordering::Relaxed)
+    }
+    fn new(location: String, other_location: String, verbose: bool) -> Self {
+        Self {
+            current_state: Arc::new(AtomicU8::new(State::NotReady.into())),
+            other_location,
+            location,
+            verbose,
+        }
+    }
+    async fn get_state(
+        from_location: String,
+        assume_on_no_response: State,
+        verbose: bool,
+    ) -> State {
+        let mut stream = match TcpStream::connect(from_location.as_str()).await {
+            Ok(stream) => stream,
+            Err(_) => return assume_on_no_response,
+        };
+        let mut buffer = Vec::new();
+        stream.read_to_end(&mut buffer).await.expect("Failed to read from peer?");
+        let option_other_state = serde_json::from_slice(&buffer).ok().map(|sm: StateMessage| sm.state);
+        if let Some(other_state) = option_other_state {
+            log_polling!(
+                verbose,
+                "{} reports state: {:?}",
+                from_location,
+                other_state
+            );
+            other_state
+        } else {
+            log_polling!(
+                verbose,
+                "{} unavailable, assuming state: {:?}",
+                from_location,
+                assume_on_no_response
+            );
+            assume_on_no_response
+        }
+    }
+    fn poll(
```

Comment:
nit: Poll tends to have pretty specific async context things, at least for me personally, it's a little confusing seeing a poll implementation that isn't using the standard poll interface. Maybe rename this?